src/main/fc/init.c

```diff
@@ -445,6 +444,16 @@ void init(void)
 
     systemState |= SYSTEM_STATE_CONFIG_LOADED;
 
+#ifdef USE_SDCARD
+    // Ensure the SD card is initialised before the USB MSC starts to avoid a race condition
+#if !defined(CONFIG_IN_SDCARD) && defined(STM32H7) && defined(USE_SDCARD_SDIO) // H7 only for now, likely should be applied to F4/F7 too
```

Comment:
Not sure why this change is needed on F7 - MSC for SD card seems to be working just fine for F4 / F7 with the current initialisation order?

I am a bit wary of changing the order, as there are a lot of external dependencies (like ESC initialisation) that rely on the timing of things happening after power up - so I expect that if this is changed there will be fallout for some of these.

Reply:
This isn't needed if the config is stored on the SD card as in that case sdCardAndFSInit() is called early, prior to calling mscStart(). With the config in FLASH, without this change (and the corresponding one you comment on below) the FC hangs when MSC tries to mount the SD card before it's initialised. Rather than risk MSC timeouts I figured it was best to avoid this race condition. Such timeouts appear to have been a problem before with onboard flash judging by the comment.

```
        // If the blackbox device is onboard flash, then initialize and scan
        // it to identify the log files *before* starting the USB device to
        // prevent timeouts of the mass storage device.
```

I believe this is low risk as all this does it ensure the SD card is initialised before MSC irrespective of if the config is stored on the SD card. The SD_INIT_ATTEMPTED flag is already there to accommodate the SD being initialised at varying time depending on config. The startup timing with this change is then close to if the config was in SD.

Tested on both an SPRACINGH7ZERO and a MATEKF722SE, the only FCs I have with SD.

Reply:
> With the config in FLASH, without this change the FC hangs when MSC tries to mount the SD card before it's initialised.

This sounds a bit strange, seeing that all MCU types except for H7 **only** support the config in flash, and MSC seems to be working fine there. Any idea why this is?

But if there really is a problem with the initialisation order when config in flash is used then this should be changed for all MCU types and not only H7.

Reply:
The issue is because we're using the HAL SDIO driver, not the SPI driver used by other processors. I'll test this on my MATEKF722SE and report back.

Reply:
Actually as we're only moving the initialisation for targets with USE_SDCARD_SDIO defined, this won't affect F7/F7 targets except for the WORMFC and Nucleo targets. So testing SPI SD Card access on my MATEKF722SE to test such a change won't prove anything. I'm not able to test F4 SDIO support so I'm inclined to leave well alone. That said, I can't see SDIO_GPIO_Init() being called anywhere else, so does SDIO actually work on any F4/F7 card?

The issue here is that we're using the SDIO, not SPI driver anyway, so I think it's entirely reasonable to expect different race conditions etc.

Reply:
- the WORMFC has got SDIO working on F4, and I've tested SDIO on a F745 test board, so this works as well. Not sure about the F722, there were bugs in the STM SDIO library for this one a while back;
- since (almost) all F4 / F7 targets are using Unified Target binaries that include support for SDIO, and since H7 will be moved to a Unified Target (at least for H743) at some point, the code needs to work for SDIO and SPI attached SD cards.

Reply:
Should we go full monty here and remove the conditional on STM32H7? Unless we _know_ that H7 needs a special case for control flow here, I'd be for standardising here.

Reply:
I agree, but I can't test this. Are you able to do so? If you have an F4/F7 FC with SDIO this could be tested. I see that Joshua Bardwell is launching an F7 with SD in preference to onboard FLASH. I'll ask if this has SPI or SDIO connectivity and blag one if the latter.

Reply:
I don't have access to my Nucleo F745 that I used to test SDIO on F7 at the moment. Maybe @jflyper has got a test jig and is able to test if this can be enabled for F7 (or F4) with SDIO?

Reply:
Should we put that in another PR?

Reply:
I wouldn't want to wait with changing this much longer so we get it exposed to users of the Unified Targets (that support SDIO), so my hunch would be to change it now and see if we get feedback.

Reply:
Done